[CEP 27] A simple Sigstore predicate (#112)
conda/ceps: cep-sigstore-predicate.md

# CEP - Standardizing a publish attestation for the conda ecosystem

<table>
<tr><td> Title </td><td> Standardizing a publish attestation for the conda ecosystem </td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
<tr><td> Created </td><td> Feb 18, 2025 </td></tr>

> <tr><td> Updated </td><td> Feb 18, 2025</td></tr>

Reviewer: nitpick: updated should be updated.

     claimant identity. The certificate issued by Fulcio is, in turn auditable
     via [RFC 6962] Certificate Transparency (CT) logs.

    - Alice signs her attestation with her ephemeral key, and distributes a
      "bundle" containing both her attestation and her signing certificate.

    - Instead of establishing trust a long-lived key from Alice, Bob establishes
      trust in Alice's identity.

    - Bob can verify the attestation's signature against Alice's emphemeral key,
      which in turn can be verified as authentically Alice's via the Fulcio-
      issued certificate.

    With this flow, neither Alice nor Bob needs to maintain long-lived signing
    or verifying keyrings, in turn reducing the attacker surface for key
    compromise.

  Another key misuse-resistance contribution within Sigstore is *machine
  identities*. A machine identity behaves similarly to a human identity
  (Alice or Bob), but identifies a machine instead of a human. For example,
  `github.com/example/example/.github/workflows/release.yml@refs/tags/v1.2.3`
  could be the machine identity of a GitHub Actions workflow that ran from
  `release.yml` within `example/example` against the `v1.2.3` tag.

> ## Motivation

Reviewer: The motivation still seems *mostly* catered toward "attestations in general" and makes very little mention that this is "attesting about the publication metadata".  I wonder if this could be slightly reworded to emphasize the actual goals of `publish/v1`

Author: Any suggestions how to reword it?

   Sigstore bundle.
1. The verifier performs a standard Sigstore verification process against
   the bundle, using Alice's identity (or machine identity) as the
   signing identity. This process produces a verified in-toto statement.

   This step requires the verifier to establish trust in the identity
   being verified against.

   Exact mechanisms for establishing this trust are
   outside the scope of this CEP; see [Future work](#future-work).

1. The verifier checks the in-toto statement for consistency against their
   ground truth:

   - The `predicateType` field **MUST** be `https://schemas.conda.org/attestations/publish/v1`.
   - The `subject[0].name` field **MUST** match the filename of the conda package.
   - The `subject[0].digest` field **MUST** match the SHA256 hash of the conda
     package.
   - The `predicate.targetChannel` field **SHOULD** match the channel that
     the package was retrieved from, if `predicate` is present. However, the
     verifier **MAY** choose to allow a channel mismatch, e.g. if the known
     context is a mirroring context (where the conda package was originally
     published to a different channel, but is now being consumed from

>      a mirror).

Reviewer: Possibly not worth addressing in this version of the spec, however I imagine a scenario:

User creates a mirror that pulls in a few specific packages or channels together. User *wants* to validate publish attestations against all the packages as they use them (not at mirror time) -- how can a user identify that YES this package listed as published to `cf-staging`, and this other package to `somesecretorg` are valid, unless they blindly trust the set of `[null, cf-staging, somesecretorg]` to be valid for any package they use, regardless of where it truly came from.

# CEP - Standardizing a publish attestation for the conda ecosystem

<table>
<tr><td> Title </td><td> Standardizing a publish attestation for the conda ecosystem </td>
<tr><td> Status </td><td> Proposed </td></tr>

> <tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>

Author: Added!

<tr><td> Created </td><td> Feb 18, 2025 </td></tr>
<tr><td> Updated </td><td> Jun 18, 2025</td></tr>

> <tr><td> Discussion </td><td>  </td></tr>

Reviewer 2: ```suggestion
<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/112 </td></tr>
```

<tr><td> Implementation </td><td>  </td></tr>
</table>

## Abstract

This CEP proposes a standard attestation layout for the conda ecosystem.
This attestation layout is based on the [in-toto] framework
and will enable further integration with signing schemes like
[Sigstore].

> 

Reviewer 3: As a stylistic note, we should add something like this, since it looks like we're using "MUST" in the RFC2119-way:

> ## Specification
>
> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119][RFC2119] when, and only when, they appear in all capitals, as shown here.
>
> More specifically, violations of a MUST or MUST NOT rule MUST result in an error. Violations of the rules specified by any of the other all-capital terms MAY result in a warning, at discretion of the implementation.

    - Alice signs a machine readable statement equivalent to the following
      English sentence, producing her attestation:

        > Alice published the `widgets` package at version v1.2.3 with
        > hash `sha256:abcd...` to the `conda-forge` channel.

    - Bob establishes trust in Alice's public key.
    - Bob can verify the attestation's signature against Alice's public key,
      giving him confidence that the statement is true.
    - Correspondingly, Bob can reject any statement for `widgets` that is not
      signed by Alice's public key.

- [in-toto] is a framework and standard for defining attestations.

    - Within in-toto, an attestation's statement is composed of a
      **subject** and a **predicate**. The subject is the resource
      (or resources) being attested to, and the predicate is a
      an arbitrary collection of metadata about the subject.
      The predicate is identified by a **predicate type**,
      which defines the predicate's expected schema.

- [Sigstore] is a project that enables misuse-resistant software signing
  and verification via short-lived certificates and a tamper-evident log.

>   Sigstore composes with attestation frameworks like in-toto to provide

Reviewer 4: fixed, changed to `Sigstore uses attestation frameworks like in-toto...`

  transparency and misuse-resistance properties on top of the integrity
  and authenticity properties of attestations.

>    One of Sigstore's major misuse-resistance contributions is

Author: Fixed.

   the use of *ephemeral keys* for signing. Modifying the example above:

   - Instead of maintaining a long-lived signing key, Alice generates an
     *ephemeral key* and binds it to her *identity*
     ("`alice@trustme.example.com`").

     This binding is done via a certificate issued by [Fulcio], which verifies a
     *proof of possession* (such as from [OpenID Connect]) from Alice for her
     claimant identity. The certificate issued by Fulcio is, in turn auditable

>      via [RFC 6962] Certificate Transparency (CT) logs.

Reviewer 5: That's whatever identity Alice actually claims, e.g. `alice@example.com` or `https://github.com/alice` or `https://github.com/alice/example/.github/workflows/publish.yml` (the latter being a machine identity for a CI/CD workflow instead of a human identity).

I could probably clarify that by dropping "claimant," since in context these are the only kinds of relevant identities 🙂

    - Alice signs her attestation with her ephemeral key, and distributes a
      "bundle" containing both her attestation and her signing certificate.

    - Instead of establishing trust a long-lived key from Alice, Bob establishes
      trust in Alice's identity.

    - Bob can verify the attestation's signature against Alice's emphemeral key,
      which in turn can be verified as authentically Alice's via the Fulcio-
      issued certificate.

    With this flow, neither Alice nor Bob needs to maintain long-lived signing
    or verifying keyrings, in turn reducing the attacker surface for key

>     compromise.

Reviewer 4: fixed

  Another key misuse-resistance contribution within Sigstore is *machine
  identities*. A machine identity behaves similarly to a human identity
  (Alice or Bob), but identifies a machine instead of a human. For example,
  `github.com/example/example/.github/workflows/release.yml@refs/tags/v1.2.3`
  could be the machine identity of a GitHub Actions workflow that ran from
  `release.yml` within `example/example` against the `v1.2.3` tag.

## Motivation

The conda ecosystem contains metadata that answers the following questions,
in part or in full:

* _Who_ (or _what_) published this package?
* _What_ is the package's hash?
* _Where_ was this package _published from_, and where _to_?
* _When_ was this package published?

However, this metadata is not currently **cryptographically verifiable**:
the consuming party must either trust it as presented, or verify it manually

> against independent sources of truth (such as a project's release history).

Reviewer 5: Good catch -- this should probably be "authenticatable" or similar, since that's the relevant "better than integrity" property.

(The underlying argument here is that a hash is "verifiable" in the sense that you can cross-check it, but you don't have an a priori reason to believe that hash is authentic. With a signature you have that authenticity, modulo trust in a key/identity.)

> against independent sources of truth (such as a project's release history).

Reviewer 4: fixed, changed to:
```
However, the authenticity of this metadata is not currently **cryptographically verifiable**:
```

Attestations that present this metadata in a cryptographically
verifiable manner are desirable for a number of reasons:

* Package maintainers wish to demonstrate the integrity and authenticity
  of their package uploads;
* Individual downstream users wish to verify the integrity and authenticity of
  packages they consume, without placing additional trust in the
  channel or channel's hosting server;
* Attestations change the sophistication and risk profile for attackers in
  defenders' favor: the attacker must be sufficiently sophisticated
  to access private key material, *and* have a risk tolerance profile that
  accepts exposure via auditable transparency logs.

More broadly, attestation schemes like the one proposed in this CEP have
seen adoption in similar and related ecosystems:

* Python (PyPA/PyPI): [PEP 740] and [PyPI - Attestations]
* NodeJS (npm): [npm - Generating provenance statements]
* Ruby (RubyGems): [rubygems/release-gem]

## Specification

### Attestation format

This CEP proposes the following attestation statement layout, using the
[in-toto Statement schema]:

> - `predicateType` **MUST** be `https://schemas.conda.org/attestations/publish/v1`

Reviewer 2: Sorry if this is an obvious question, but what are the contents of https://schemas.conda.org/attestations/publish/v1? Or is it just a string identifying _this_ attestation format? i.e. is it empty?

Reviewer 2: Let's go for the JSON schema as Jannis suggested, and let's open the relevant PR in https://github.com/conda/schemas to provide the file.

- `subject` **MUST** be a single [`ResourceDescriptor`], with the following
  constraints:
    - `subject[0].name` **MUST** be the full filename of the conda package
      that will be part of the `repodata.json` and under which it will appear on
      the server.
    - `subject[0].digest` **MUST** be a [`DigestSet`], and it **MUST** contain
      a single `sha256` entry with the SHA256 hash of the conda package.
- `predicate` **MAY** be present. If present and not `null`, it **MUST** be a
  JSON object with the following fields:
    - `targetChannel` **MUST** be a string, indicating where the package
      is being uploaded to. This field **MUST** be a valid URL with no
      trailing slashes.

An example of a compliant statement is provided below:

```json
{
    "_type": "https://in-toto.io/Statement/v1",
    "subject": [{
        "name": "file-name-0.0.1-h123456_5.conda",
        "digest": {"sha256": "01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b"},
    }],

>     "predicateType": "https://schemas.conda.org/attestations/publish/v1",

Reviewer 4: fixed

    "predicate": {
        "targetChannel": "https://prefix.dev/conda-forge",
    }
}
```

### Signing and distributing

This CEP recommends the following signing process:

1. The signer (i.e. Alice or Alice's trusted machine identity) uses a
   [Sigstore]-compatible client to generate an ephemeral keypair and bind it to
   their identity via a public certificate.
2. The signer generates an in-toto statement as described above, and
   produces an attestation by signing that statement with their ephemeral
   private key.
3. The signer uploads their attestation to the Sigstore transparency log

>    as a [DSSE] envelope.

Reviewer 5: Yeah, in context this was intended to be the "public good" instance of Sigstore.

> what are the downsides of using a 3rd party service for trust?

In a traditional PKI, I think the biggest downside is unbounded trust extension: a third-party issuer can "go rogue" and do immense damage to the parties that depend on it. Sigstore mitigates that (successfully IMO) with multiple layers of cryptographic transparency (Fulcio's CT log, Rekor for signature transparency).

Apart from that, I think the single biggest downside/risks are stability and longevity: the assumption is that Sigstore (1) remains available, (2) has as-good-or-better uptime characteristics than its downstreams, and (3) doesn't disappear in the next few years. It's my personal opinion that Sigstore satisfies these, which is why I felt comfortable proposing it as a third-party service within PyPI. Inductively GitHub seems to feel the same way, since their public attestation features are built on top of Sigstore's public good instance.

4. The signer produces a [Sigstore bundle] containing their certificate,
   attestation, and transparency log inclusion proof.

Each of these steps is performed transparently by a Sigstore client like
[sigstore-python], except for step (2) as it concerns the specific

> layout of the signed-over statement.

Reviewer 5: Ah, this is poorly phrased -- I was trying to split between the things that sigstore-python can do _internally_ versus the things a user needs to feed into it. Everything in the above steps is done by _interacting_ with sigstore-python's APIs or CLI, it's just that (2) is where the user tells sigstore-python _what_ to sign 🙂

> layout of the signed-over statement.

Reviewer 4: fixed, rephrased to:
```
Each of these steps is performed transparently by a Sigstore client like
[sigstore-python], except for the generation of the in-toto statement in
step (2), which should be handled externally and then fed into the Sigstore
client for signing.
```

   Exact mechanisms for establishing this trust are
   outside the scope of this CEP; see [Future work](#future-work).

1. The verifier checks the in-toto statement for consistency against their
   ground truth:

   - The `predicateType` field **MUST** be `https://schemas.conda.org/attestations/publish/v1`.
   - The `subject[0].name` field **MUST** match the filename of the conda package.
   - The `subject[0].digest` field **MUST** match the SHA256 hash of the conda
     package.
   - The `predicate.targetChannel` field **SHOULD** match the channel that
     the package was retrieved from, if `predicate` is present. However, the
     verifier **MAY** choose to allow a channel mismatch, e.g. if the known
     context is a mirroring context (where the conda package was originally
     published to a different channel, but is now being consumed from
     a mirror).

At the end of this process, the verifier is confident in the following facts:

- The package was published by the signer (Alice or Alice's machine identity).
  - If the publisher is a machine identity, this further establishes source
    provenance via the machine identity's claims. See [Sigstore OID information]

>     for additional information on these claims.

Reviewer 4: I think "build provenance" might be more appropriate? The machine identity claims are about the build stage (which source revision was used to build, what environment was used to execute the build, what workflow defined the build, etc), rather than about the source.
At least following the definitions in https://slsa.dev/spec/draft/provenance, I think `build provenance` makes more sense.

- The package is authentic and integral modulo trust in the signer.

This CEP suggests that attestation verification be performed by
both clients (i.e. package installers retrieving packages from
attestation-aware channels) and servers (i.e. attestation-aware channels).

Servers **SHOULD** perform the same verification process as clients,
with the qualification that the server's trust in the signing identity
is established latently via the server's publishing and upload mechanism.
For example, if the server supports [Trusted Publishing], then the package's
attestation should be verified against the set of Trusted Publisher identities

> for that package.

Reviewer 5: I see no reason to have it not be a `MUST` -- the argument for having it be a `SHOULD` is that it _could_ be useful to convey attestations via a server without being able to verify them directly. However we intentionally didn't want to encourage that on PyPI with PEP 740 and so the same likely doesn't apply here 🙂

> for that package.

Reviewer 4: fixed, `SHOULD` -> `MUST`

## Security Model

This section provides a high-level security model for the scheme proposed
in this CEP.

### Unforgeability of provenance

Traditional package distribution schemes depend on untrusted metadata
for provenance: the package itself describes its source repository,
its maintainers, etc.

This allows an attacker to forge a package's metadata such that a typical
downstream unduly trusts the package. For example, a typical downstream
may observe that the package lists a "trustworthy" GitHub repository
or organization as its source, and therefore trust the package without
confirming that its contents actually reflect those of the repository.
This is compounded when the *distribution format* (such as a conda package

> or Python wheel) does not directly match the claimant repository's

Reviewer 4: fixed, reworded to:

```
This is compounded when the *distribution format* (such as a conda package
or Python wheel) does not directly match the source layout (such as the
contents of a GitHub repository), making direct comparison nontrivial.
```

source layout, making direct comparison nontrivial.

This CEP's design introduces an *unforgeability* property: a package
that claims a piece of untrusted metadata can be verified by cross-checking
that metadata against the package's attestation.

An attacker cannot forge or spoof this check, modulo their ability
to compromise the authentic signing identity itself. This is effectively
a stronger unforgeability property than traditional signing, as
the signature binds not only the package's contents, but also the package's

> provenance.

Reviewer 5: Yeah, I could probably remove this -- the original intent was to note that the scheme has the same existential unforgeability property (EUF) as its underlying digital signature scheme, but the mention of the domain binding confuses things significantly there.

(The argument that the attacker cannot forge the check is standard under EUF, which is assumed via Sigstore's choice of supported signature schemes. But the part about being "stronger" than traditional signing is a red herring that I'll remove.)

> provenance.

Reviewer 4: fixed, removed the part about being stronger:
```
An attacker cannot forge or spoof this check, modulo their ability
to compromise the authentic signing identity itself.
```

### Transparency and auditability

Traditional signing schemes provide integrity and authenticity
modulo trust in a signing identity.

This is a strong property, but not a perfect one: an attacker who *does* manage

> to compromise a signing identity can mount a *targeted* attack, wherein a
> the general public observes only legitimate artifacts and signatures while

Reviewer 6: ```suggestion
to compromise a signing identity can mount a *targeted* attack, wherein
the general public observes only legitimate artifacts and signatures while
```

the victim receives a malicious artifact with a valid signature from the
compromised signing identity. In effect, this makes it possible for the
attacker to maintain their stealth during a *targeted* attack, since
*untargeted* parties are not made aware of the malicious artifact or
its signature.

This CEP's design introduces a *transparency* property: attestations
are not considered valid unless they are included in a publicly
auditable, append-only transparency log. This log is made up of *entries*,
which bind over the attestation itself (including the subject),
the attestation's signature, as well as the attestation's signing identity.

Because a valid transparency log inclusion proof is a requirement during
verification, an attacker who compromises a signing identity *cannot*
perform a targeted attack without making the attack itself publicly

> discoverable.

Reviewer 5: > Can you say something about how this added transparency influences the ability to defend against targeted attacks?

Yeah, can do -- the idea is that in the context of targeted attacks, the victimized party (i.e. the one the attacker is impersonating) can monitor for uses of their identity in real time, meaning that they can respond proactively to the compromise instead of waiting for the first report of compromise from a downstream. I'll fill that in.

> discoverable.

Reviewer 4: Fixed, added more about how the transparency log helps detecting and responding to compromises:

```
Because a valid transparency log inclusion proof is a requirement during
verification, an attacker who compromises a signing identity *cannot*
perform a targeted attack without making the attack itself publicly
discoverable. This means the log can be monitored in real time for
uses of identities, allowing users to proactively detect and respond
to an identity being compromised, rather than having to wait for the
compromise being detected downstream and reported.
```

   analogous to the key distribution problem in traditional PKI systems,
   albeit without the complications and operational challenges associated
   with long-lived key material.

    One potential mechanism for identity trust distribution is a TOFU (trust on
    first use) scheme with an attestation-aware conda channel, where package
    names are "locked" to attesting identities on first use, with subsequent
    updates being verified against that identity. This scheme requires a
    suitable lockfile format; [PEP 751] and its
    `[[package.attestation-identities]]` table may serve as prior art
    for a similar approach in the conda ecosystem.

2. This CEP does not specify a distribution mechanism for attestations
   (i.e., Sigstore bundles containing attestations).

    One potential distribution mechanism is to have attestation-aware
    conda channels distribute each package's attestations alongside
    the package, or via a similarly discoverable channel-side API.
    Prior art for this type of distribution mechanism can be found in
    the PyPI and RubyGems ecosystems, e.g. [PyPI's Integrity API].

3. This CEP specifies an single initial in-toto predicate

>    (`https://schemas.conda.org/attestations/publish/v1`), which conveys

Reviewer 4: fixed

# CEP - Standardizing the v1 predicate for sigstore attestations

<table>
<tr><td> Title </td><td> Standardizing the v1 predicate for sigstore attestations </td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
<tr><td> Created </td><td> Feb 18, 2025 </td></tr>
<tr><td> Updated </td><td> Feb 18, 2025</td></tr>
<tr><td> Discussion </td><td>  </td></tr>
<tr><td> Implementation </td><td>  </td></tr>
</table>

## Abstract

We want to standardize attestations for the conda ecosystem.

### Sigstore Attestations

> Sigstore attestations are cryptographic statements about software artifacts that provide:

Reviewer 3: Does "SigStore attestation" here mean the SigStore bundle itself or some other attestation payload defined elsewhere, wrapped as the SigStore verification material? To reduce confusion for someone reading those specs, we should use the same [language that SigStore does](https://docs.sigstore.dev/about/bundle/).

- Authenticity: Proof of who created/signed the artifact

> - Integrity: Verification that the artifact hasn't been tampered with

Reviewer 3: SigStore cannot directly make any integrity claims; integrity is tied to the sturcture and contents of the verification material, and those are defined outside the SigStore framework.

- Transparency: Public record of signatures in a tamper-evident log

### Key Components

> - Predicates: JSON documents containing metadata about the signing event, using the `in-toto` format

Reviewer 3: Strictly speaking, predicates are not part of the metadata about a SigStore signing event. "Predicates" may exist as part of the SigStore verification material, if said material is a DSSE envelope.

Author: I think this is outdated after the rewrite.

- Signatures: Cryptographic proofs made using ephemeral keys
- Rekor: A tamper-evident log that stores attestations

> - Fulcio: A certificate authority that issues short-lived certificates

Reviewer 3: Is the claim of this CEP to make `rekor` and `falcio` the only acceptable transparency log and certificate authority for the conda ecosystem?

In this document, we want to standardize the sigstore predicate for conda packages. The bundle format to be used for sigstore attestations is the `v0.3` bundle format.

## Specification

The in-toto predicate should contain the following fields:

```json
{

>     "_type": "https://in-toto.io/Statement/v0.1",

Reviewer: This comment is still outstanding

>     "_type": "https://in-toto.io/Statement/v0.1",

Author: Fixed.

>     "subject": [{

Reviewer 3: Is the `subject` field constrained to one and only one package? Or are there circumstances in which multiple packages can be covered by a single attestation?

Reviewer: Still an open question (per the current spec, `subject` is one-package only), but worth calling out that the comment is un-answered.

Reviewer 5: Thanks, I forgot to answer directly here: this CEP as-drafted proposed only a single package subject at a time, since allowing multiple subjects introduces (IMO) a degree of malleability that makes downstream usage harder.

To be precise: if an attestation is allowed to contain multiple subjects, a malicious party could do something like:

```
subjects:
  authentic-0.0.1-etc.conda
  inauthentic-0.0.1-etc.conda
```

...where the `authentic` package identity is one they're authorized (via the server) to publish for, while `inauthentic` is some other package they aren't authorized for. This is turn would make it easy for the server to accidentally accept "mixed" attestations and re-broadcast them as if they're "fully" verified.

This is mentioned under the `Verifying` section as well:

> Servers **SHOULD** perform the same verification process as clients,
with the qualification that the server's trust in the signing identity
is established latently via the server's publishing and upload mechanism.
For example, if the server supports [Trusted Publishing], then the package's
attestation should be verified against the set of Trusted Publisher identities
for that package.

        "name": "file-name-0.0.1-h123456_5.conda",

>         "digest": {"sha256": "..."}, ...

Author: It implies that there may be more than one subject. You can find the spec here https://in-toto.io/Statement/v0.1

    }],
    // Schema URL

>     "predicateType": "https://schemas.conda.org/predicate-v1.json",

Author: What would be your proposal? `predicate-base.json`? or `simplePredicate-v1.json`?

>     "predicateType": "https://schemas.conda.org/predicate-v1.json",

Reviewer 3: `predicate-{thing we're attesting to}-v{version}.json`? IMO, the name of the attestation schema should reflect the threat models and associated security claims, which I'm not sure are clearly stated in this CEP.

As an example, PyPI [currently] supports [two different attestation predicate](https://docs.pypi.org/attestations/#supported-attestations), one named "publish" (making claims about who put a package on PyPI) and another named "provenance" (making claims about how the package on PyPI was built). We should follow that pattern.

>     "predicateType": "https://schemas.conda.org/predicate-v1.json",

Reviewer 3: Also, if we're going to follow [in-toto parsing rules](https://github.com/in-toto/attestation/blob/main/spec/v1/README.md#parsing-rules), it might actually be easier if the `predicateType` URI is not actually a file, since said rules require that the URI only expose the major version number. If we use files, we'll likely have to do some symlinking or other magic to make sure we can have schemas for v1.0, v1.1, etc.

>     "predicateType": "https://schemas.conda.org/predicate-v1.json",

Reviewer: This is now `.../publish/v1` which feels like it resolves this comment

    "predicate": {
        // Canonical URL of the target channel
        "targetChannel": "https://prefix.dev/conda-forge",
    }
}
```

The `subject` field is already defined in the in-toto specification and contains the name of the package and its digest.

> For conda packages a SHA256 hash MUST be used.

Reviewer 3: Suggest rewording to:
> For conda packages, the digest MUST include the package's SHA256.

The specification should be clear if only the SHA256 can be included, or if optionally, other package checksums can be included as well.

Author: Reworded to: 

>     - `subject[0].digest` **MUST** be a [`DigestSet`], and it **MUST** contain
      a single `sha256` entry with the SHA256 hash of the conda package.

The subject MUST be the full filename of the conda package that will be part of the repodata.json and under which it will appear on the server.

The `predicateType` field is used to specify the schema of the predicate. The `predicate` field contains the actual predicate data.
We propose to publish a schema to validate the `predicate` field. The schema will be available at `https://schemas.conda.org/predicate-v1.json`.

> The predicate MUST contain the `targetChannel` field, to indicate where the package is being uploaded to. This field MUST be validated by the receiving server. The channel MUST be in canonical form (full URL, no trailing slashes).

Reviewer 7: ```suggestion
The predicate MUST contain the `targetChannel` field, to indicate where the package is being uploaded to. This field's value MAY be validated by the receiving server, but such validation is not possible in general and thus not required. The `targetChannel` MUST be in canonical form (full URL, no trailing slashes).
```

Author: @beckermr how would you feel about making it an optional field? I think if the field exists, it should be validated.

Reviewer 7: I thought the predicate had to be non-empty in order to sign at all?

Reviewer 7: Ah ok. Yes. I'm happy to have it be optional, but if it is present it must be validated. I'll change my review.

> The predicate MUST contain the `targetChannel` field, to indicate where the package is being uploaded to. This field MUST be validated by the receiving server. The channel MUST be in canonical form (full URL, no trailing slashes).

Reviewer 3: What does the predicate material attest to? Target channel makes it seem like what we want to attest to is some like PyPI Trusted Publishing, but that is not at all clear from the content of this proposed CEP.

If what we want to attest to are claims about the security of build process (a la SLSA), this predicate specification is definitely not sufficient.

Reviewer 3: Some other notes:

> ...[channel where the] package is being uploaded to...

We might want to clarify or refine that clause. I might be wrong, but a strict reading of that could mean that channel value for conda-forge should be "cf-staging", not "conda-forge". (Since packages are initially uploaded to "cf-staging" and then _copied_ over to "conda-forge".) 

> channel MUST be in canonical form

We need some reference defining what "canonical form" means. I'm not sure there's consensus of a "canonical" channel in the conda-verse, since packages are perhaps surprisingly mobile across channels.

Author: I wouldn't want to generalize from what `conda-forge` does. We could very well special case how we validate packages for conda-forge based on the workarounds that are currently applied, but I don't think it's a _nice_ situation that conda-forge is in from which we should spec out how we want to deal with sigstore. 

conda-forge could decide to upload to a S3 bucket first, and then upload to the conda-forge channel, with another Github Action that also makes OIDC trusted publishing work, as an example.

As such, I don't think this needs any clarification.

And yes, for the canonical form, I propose to use the FULL URL for the channel without trailing slash, e.g. `https://anaconda.org/conda-forge`. This has nothing to do with packages being mobile, which is another concern that we could solve in a number of ways (not relevant here).

> 

Reviewer 7: ```suggestion

On the server, the certificate MUST be tested against the Trusted Publisher used to upload the certificate to establish a chain of trust.

```

Reviewer 8: Is it defined what a trusted publisher is? Perhaps we could rephrase tonsomething like:, the server must validate that the signee is a trusted entity. 

What a trusted entity is, is left up to the implementation.

Reviewer 7: I copied this out of the discussion part since it previously read

> On the server, the certificate should be tested against the Trusted Publisher used to upload the certificate to establish a chain of trust.

I assumed whoever wrote it knew what it meant and when they said "should" they actually meant MUST. 

@wolfv can you elaborate?

Author: I think we should change it to MUST to have the same guarantees as PyPI.

Reviewer: Obsolete Comment?

## Discussion

> This predicate adds basic verifiable facts about the package. It will tie the producer of the package to the target channel.
> This is similar to what PyPI has implemented with the [PyPI publish attestation](https://docs.pypi.org/attestations/publish/v1/). Since there is no single authoritative index in the Conda world, we add the `targetChannel` field to reach parity.

Reviewer 7: ```suggestion
This predicate ties the producer of the package to the initial target channel for that package. The initial target channel information may become out of date as channels, their locations, names, etc. evolve. This implementation is similar to the [PyPI publish attestation](https://docs.pypi.org/attestations/publish/v1/), but includes the `targetChannel` since there is no single authoritative `conda` package index.
```

> On the server, the certificate should be tested against the Trusted Publisher used to upload the certificate to establish a chain of trust.

Reviewer 7: ```suggestion
```

> On the server, the certificate should be tested against the Trusted Publisher used to upload the certificate to establish a chain of trust.

Reviewer 3: It's not clear to me what which server is meant here, nor what chain of trust is being established. (Missing threat model in the CEP.)

Reviewer: The current CEP still seems to be missing a sense of "what threat will/may this protect against" - It would be great to flesh that out a bit, rather than simply the "what are we specifying in the attestation"
